### PR TITLE
Add NextMillenium bidder

### PR DESCRIPTION
**Pull request by yevhenii-viktorov**
Original ticket: https://github.com/prebid/prebid-server/pull/2016/
Additional changes: https://github.com/prebid/prebid-server/pull/2061

### CODE REVIEW

File: src/test/resources/org/prebid/server/it/test-application.properties [hunk 1]
```diff
@@ -157,6 +157,8 @@ adapters.mobilefuse.enabled=true
 adapters.mobilefuse.endpoint=http://localhost:8090/mobilefuse-exchange/
 adapters.nanointeractive.enabled=true
 adapters.nanointeractive.endpoint=http://localhost:8090/nanointeractive-exchange/
+adapters.nextmillenium.enabled=true
+adapters.nextmillenium.endpoint=http://localhost:8090/nextmillenium-exchange/
```
Comment (snahornyi) [inline src/test/resources/org/prebid/server/it/test-application.properties:161]: Minor. Remove redundant `/` make value to be `/nextmillenium-exchange`

File: src/test/java/org/prebid/server/it/NextMilleniumTest.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import static com.github.tomakehurst.wiremock.client.WireMock.urlPathEqualTo;
+import static java.util.Collections.singletonList;
+
+@RunWith(SpringRunner.class)
+public class NextMilleniumTest extends IntegrationTest {
+
+    @Test
+    public void openrtb2AuctionShouldRespondWithBidsFromNextMillenium() throws IOException, JSONException {
+        // given
+        WIRE_MOCK_RULE.stubFor(post(urlPathEqualTo("/nextmillenium-exchange/"))
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/it/NextMilleniumTest.java:24]: Minor. Remove redundant `/` make value to be `/nextmillenium-exchange`

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/nextmillenium/ExtImpNextMillenium.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+
+/**
+ * Defines the contract for bidRequest.imp[i].ext.nextmillenium
+ */
+@AllArgsConstructor(staticName = "of")
+@Value
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/nextmillenium/ExtImpNextMillenium.java:11]: @Value(staticConstructor = "of")

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/nextmillenium/ExtImpNextMillenium.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+public class ExtImpNextMillenium {
+
+    @JsonProperty("placement_id")
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/nextmillenium/ExtImpNextMillenium.java:14]: Unnecessary `@JsonProperty` default strategy is already snake case

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/nextmillenium/ExtImpNextMillenium.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package org.prebid.server.proto.openrtb.ext.request.nextmillenium;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+
+/**
+ * Defines the contract for bidRequest.imp[i].ext.nextmillenium
+ */
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/nextmillenium/ExtImpNextMillenium.java:9]: We are going to get rid of useless comments, pls remove it

File: src/main/java/org/prebid/server/spring/config/bidder/NextMilleniumConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    @ConfigurationProperties("adapters.nextmillenium")
+    BidderConfigurationProperties configurationProperties() {
+        return new BidderConfigurationProperties();
+    }
+
+    @Bean
+    BidderDeps nextmilleniumBidderDeps(BidderConfigurationProperties nextmilleniumConfigurationProperties,
+                                       @NotBlank @Value("${external-url}") String externalUrl,
+                                       JacksonMapper mapper) {
+        return BidderDepsAssembler.forBidder(BIDDER_NAME)
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/spring/config/bidder/NextMilleniumConfiguration.java:34]: NitPick. `/n` before return statement

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                                .build()))
+                .build();
+    }
+
+    @Override
+    public final Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            if (bidResponse.getSeatbid().size() == 0) {
+                return Result.of(null, null);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:118]: return Result.empty();

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+                                .storedrequest(ExtStoredRequest.of(ext.getPlacementId()))
+                                .build()))
+                .build();
+    }
+
+    @Override
+    public final Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            if (bidResponse.getSeatbid().size() == 0) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:117]: CollectionUtils.isEmpty(bidResponse.getSeatbid())

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+                        .convertValue(imp.getExt(), NEXTMILLENIUM_EXT_TYPE_REFERENCE)
+                        .getBidder());
+            } catch (IllegalArgumentException e) {
+                errors.add(BidderError.of(e.getMessage(), BidderError.Type.bad_input));
+            }
+        }
+        return impExts;
+    }
+
+    private List<ExtImpNextMillenium> errorIfEmpty(List<BidderError> errors) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:76]: BidRequest validator checks is there any imp in request, even more without imp request will not come to bidder. This method is redundant

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    @Override
+    public final Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest bidRequest) {
+        final List<Imp> imps = bidRequest.getImp();
+        final List<BidderError> errors = new ArrayList<>();
+        final List<ExtImpNextMillenium> extImps = CollectionUtils.isNotEmpty(imps)
+                ? getImpressionInfo(imps, errors)
+                : errorIfEmpty(errors);
+
+        if (errors.size() > 0) {
+            return Result.of(null, errors);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:56]: return Result.withErrors(errors);

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        } else {
+            return Result.of(makeRequests(bidRequest, extImps), Collections.emptyList());
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:58]: return Result.withValues(makeRequests(bidRequest, impExts));

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    public final Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest bidRequest) {
+        final List<Imp> imps = bidRequest.getImp();
+        final List<BidderError> errors = new ArrayList<>();
+        final List<ExtImpNextMillenium> extImps = CollectionUtils.isNotEmpty(imps)
+                ? getImpressionInfo(imps, errors)
+                : errorIfEmpty(errors);
+
+        if (errors.size() > 0) {
+            return Result.of(null, errors);
+        } else {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:57]: Else can be ommited

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @Override
+    public final Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest bidRequest) {
+        final List<Imp> imps = bidRequest.getImp();
+        final List<BidderError> errors = new ArrayList<>();
+        final List<ExtImpNextMillenium> extImps = CollectionUtils.isNotEmpty(imps)
+                ? getImpressionInfo(imps, errors)
+                : errorIfEmpty(errors);
+
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:54]: We can update this to this, and also I would suggest to extract ext resolving logic to another method
```java
        final List<BidderError> errors = new ArrayList<>();
        final List<ExtImpNextMillenium> impExts = new ArrayList<>();

        for (Imp imp : bidRequest.getImp()) {
            try {
                impExts.add(mapper.mapper()
                        .convertValue(imp.getExt(), NEXTMILLENIUM_EXT_TYPE_REFERENCE)
                        .getBidder());
            } catch (IllegalArgumentException e) {
                errors.add(BidderError.of(e.getMessage(), BidderError.Type.bad_input));
            }
        }
```

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        return httpRequests;
+    }
+
+    private HttpRequest<BidRequest> makeHttpRequest(BidRequest bidRequest) {
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(HttpUtil.headers().add(HttpUtil.X_OPENRTB_VERSION_HEADER, "2.5"))
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:96]:
```java
private static MultiMap headers() {

        return HttpUtil.headers()
                .add(HttpUtil.X_OPENRTB_VERSION_HEADER, "2.5");
    }
```

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    private List<HttpRequest<BidRequest>> makeRequests(BidRequest bidRequest, List<ExtImpNextMillenium> extImps) {
+        final List<HttpRequest<BidRequest>> httpRequests = new ArrayList<>();
+
+        for (ExtImpNextMillenium ext : extImps) {
+            httpRequests.add(makeHttpRequest(updateBidRequest(bidRequest, ext)));
+        }
+
+        return httpRequests;
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:89]: Let's update with stream approach
```java
private List<HttpRequest<BidRequest>> makeRequests(BidRequest bidRequest, List<ExtImpNextMillenium> extImps) {

        return extImps.stream()
                .map(extImp -> makeHttpRequest(updateBidRequest(bidRequest, extImp)))
                .collect(Collectors.toList());

    }
```

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(HttpUtil.headers().add(HttpUtil.X_OPENRTB_VERSION_HEADER, "2.5"))
+                .payload(bidRequest)
+                .body(mapper.encodeToBytes(bidRequest))
+                .build();
+    }
+
+    private BidRequest updateBidRequest(BidRequest bidRequest, ExtImpNextMillenium ext) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:102]: Can be static

Comment (snahornyi) [reply src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:102]: Pls put methods in call order

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static java.util.function.Function.identity;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
+
+public class NextMilleniumBidderTest extends VertxTest {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:34]: Pls add tests for present payload check

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+        return bidRequestCustomizer.apply(BidRequest.builder()
+                        .id("1500")
+                        .test(100)
+                        .imp(impCustomizers.stream()
+                                .map(NextMilleniumBidderTest::givenImp)
+                                .collect(Collectors.toList())))
+                .build();
+    }
+
+    private static BidRequest givenBidRequest(Function<Imp.ImpBuilder, Imp.ImpBuilder>... impCustomizers) {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:158]: UnaryOperator

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+        // when
+        final Result<List<BidderBid>> result = nextMilleniumBidder.makeBids(httpCall, bidRequest);
+
+        // then
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Failed to decode:");
+    }
+
+    private static BidRequest givenBidRequest(
+            Function<BidRequest.BidRequestBuilder, BidRequest.BidRequestBuilder> bidRequestCustomizer,
+            List<Function<Imp.ImpBuilder, Imp.ImpBuilder>> impCustomizers) {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:147]: UnaryOperator
Pls check similar

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java (hunk 18 above)
Comment (snahornyi) [reply src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:158]: Seems in your case one impCustomizer is enough, no need for varargs here

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public final Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest bidRequest) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<ExtImpNextMillenium> impExts = getImpExts(bidRequest, errors);
+
+        if (errors.size() > 0) {
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:52]: Pls refactor to use ternary operation:
`return errors.isEmpty() ? result1 : result2;`

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+                        .getBidder());
+            } catch (IllegalArgumentException e) {
+                errors.add(BidderError.of(e.getMessage(), BidderError.Type.bad_input));
+            }
+        }
+        return impExts;
+    }
+
+    private List<HttpRequest<BidRequest>> makeRequests(BidRequest bidRequest, List<ExtImpNextMillenium> extImps) {
+
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:75]: Redundant line.

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(headers())
+                .payload(bidRequest)
+                .body(mapper.encodeToBytes(bidRequest))
+                .build();
+    }
+
+    private static MultiMap headers() {
+
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:103]: Redundant line.

File: src/main/resources/bidder-config/nextmillenium.yaml [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+  nextmillenium:
+    endpoint: https://pbs.nextmillmedia.com/openrtb2/auction
+    meta-info:
+      maintainer-email: appdevs@nextmillennium.io
+      app-media-types:
+        - banner
+      site-media-types:
+        - banner
+      supported-vendors:
+      vendor-id: 0
```
Comment (rpanchyk) [inline src/main/resources/bidder-config/nextmillenium.yaml:11]: Should be `1060`
As listed in https://vendor-list.consensu.org/v2/archives/vendor-list-v118.json

**Review by rpanchyk — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+    private List<ExtImpNextMillenium> getImpExts(BidRequest bidRequest, List<BidderError> errors) {
+        final List<ExtImpNextMillenium> impExts = new ArrayList<>();
+
+        for (Imp imp : bidRequest.getImp()) {
+            try {
+                impExts.add(mapper.mapper()
+                        .convertValue(imp.getExt(), NEXTMILLENIUM_EXT_TYPE_REFERENCE)
+                        .getBidder());
+            } catch (IllegalArgumentException e) {
+                errors.add(BidderError.of(e.getMessage(), BidderError.Type.bad_input));
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidder.java:66]: errors.add(BidderError.badInput(e.getMessage()));

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+        // given
+        final BidRequest bidRequest = givenBidRequest(identity());
+        final HttpCall<BidRequest> httpCall = givenHttpCall(bidRequest,
+                mapper.createArrayNode().toString());
+
+        // when
+        final Result<List<BidderBid>> result = nextMilleniumBidder.makeBids(httpCall, bidRequest);
+
+        // then
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Failed to decode:");
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:149]: Pls also check error type. allSatisfy can be used

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+
+        // when
+        final Result<List<HttpRequest<BidRequest>>> result = nextMilleniumBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Cannot deserialize value of type");
+    }
+
+    @Test
+    public void makeBidsGivenCorrectBidResponseContainsCorrectBidder() throws JsonProcessingException {
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:106]: makeBidsShouldReturnBannerBidByDefault

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java [hunk 27]
```diff
@@ -0,0 +1,190 @@
...
+                .containsExactly(BidderBid.of(Bid.builder().impid("123").build(), BidType.banner, "USD"));
+    }
+
+    @Test
+    public void makeBidsWithZeroSeatBidsShouldReturnNoErrorsAndNoValues() throws JsonProcessingException {
+        // given
+        final BidRequest bidRequest = givenBidRequest(identity());
+        final HttpCall<BidRequest> httpCall = givenHttpCall(bidRequest,
+                mapper.writeValueAsString(BidResponse.builder()
+                        .seatbid(List.of())
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:127]: Pls avoid "Import with class", and also for this case use `emptyList()`

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java [hunk 28]
```diff
@@ -0,0 +1,190 @@
...
+    public void makeHttpRequestsWithInvalidImpsShouldReturnError() {
+        // given
+        final BidRequest bidRequest = givenBidRequest(impCustomizer -> impCustomizer
+                .ext(mapper.valueToTree(ExtPrebid.of(null, mapper.createArrayNode()))));
+
+        // when
+        final Result<List<HttpRequest<BidRequest>>> result = nextMilleniumBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Cannot deserialize value of type");
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:102]: Check for type also

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+                        .id("123")
+                        .ext(mapper.valueToTree(ExtPrebid.of(null,
+                                ExtImpNextMillenium.of("placement_id")))))
+                .build();
+    }
+
+    private static BidResponse givenBidResponse(UnaryOperator<Bid.BidBuilder> bidCustomizer) {
+        return BidResponse.builder()
+                .seatbid(List.of(SeatBid.builder()
+                        .bid(List.of(bidCustomizer.apply(Bid.builder()).build()))
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:178]: singletonList

File: src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java [hunk 30]
```diff
@@ -0,0 +1,190 @@
...
+        return impCustomizer.apply(Imp.builder()
+                        .id("123")
+                        .ext(mapper.valueToTree(ExtPrebid.of(null,
+                                ExtImpNextMillenium.of("placement_id")))))
+                .build();
+    }
+
+    private static BidResponse givenBidResponse(UnaryOperator<Bid.BidBuilder> bidCustomizer) {
+        return BidResponse.builder()
+                .seatbid(List.of(SeatBid.builder()
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/nextmillenium/NextMilleniumBidderTest.java:177]: singletonList

**Review by snahornyi — CHANGES_REQUESTED** [review]

**Review by snahornyi — APPROVED** [review]

**Review by rpanchyk — APPROVED** [review]

**Pull request merged**